libflux/src/semantic/types.rs

```diff
@@ -1281,4 +1345,163 @@ mod tests {
+        {
+            let sub = f.unify(call_type, &mut cons, &mut Fresher::new()).unwrap();
+            assert_eq!(
+                sub,
+                Substitution::from(maplit::hashmap! {Tvar(0) => MonoType::Int})
+            );
+            // the constraint on t0 gets removed.
+            assert_eq!(cons, maplit::hashmap! {Tvar(1) => vec![Kind::Divisible]});
+        } else {
+            panic!("the monotype under examination is not a function");
+        }
+    }
+    #[test]
+    fn unify_functions() {
+        let f = polytype(
+            "forall [t0, t1] where t0: Addable, t1: Divisible (a: t0, b: t0, ?c: t1) -> t0",
+        );
+        let g = polytype("forall [t2] where t2: Addable (a: t2, ?b: t2, c: float) -> t2");
+        if let (
+            PolyType {
+                vars: _,
+                cons: f_cons,
+                expr: MonoType::Fun(f),
+            },
+            PolyType {
+                vars: _,
+                cons: g_cons,
+                expr: MonoType::Fun(g),
+            },
+        ) = (f, g)
+        {
+            // this extends the first map with the second by generating a new one.
+            let mut cons = f_cons.into_iter().chain(g_cons).collect();
+            let sub = f.unify(*g, &mut cons, &mut Fresher::new()).unwrap();
+            assert_eq!(
+                sub,
+                Substitution::from(maplit::hashmap! {
+                    Tvar(0) => MonoType::Var(Tvar(2)),
+                    Tvar(1) => MonoType::Float,
+                })
+            );
+            // t0 is equal to t2 and t2 is Addable, so we only need one constraint on t2;
+            // t1 ended up being a float, so we do not need any kind constraint on it.
+            assert_eq!(cons, maplit::hashmap! {Tvar(2) => vec![Kind::Addable]});
+        } else {
+            panic!("the monotypes under examination are not functions");
+        }
+    }
+    #[test]
+    fn unify_higher_order_functions() {
+        let f = polytype(
+            "forall [t0, t1] where t0: Addable, t1: Divisible (a: t0, b: t0, ?c: (a: t0) -> t1) -> (d:  string) -> t0",
+        );
+        let g = polytype("forall [] (a: int, b: int, c: (a: int) -> float) -> (d: string) -> int");
+        if let (
+            PolyType {
+                vars: _,
+                cons: f_cons,
+                expr: MonoType::Fun(f),
+            },
+            PolyType {
+                vars: _,
+                cons: g_cons,
+                expr: MonoType::Fun(g),
+            },
+        ) = (f, g)
+        {
+            // this extends the first map with the second by generating a new one.
+            let mut cons = f_cons.into_iter().chain(g_cons).collect();
+            let sub = f.unify(*g, &mut cons, &mut Fresher::new()).unwrap();
+            assert_eq!(
+                sub,
+                Substitution::from(maplit::hashmap! {
+                    Tvar(0) => MonoType::Int,
+                    Tvar(1) => MonoType::Float,
```

Comment:
Ok, this fails, `t1` doesn't get constrained to `float`

Reply:
I think this is just because you forgot to unify the return types.